crates/fuel-streams-core/src/utxos/subjects.rs

```diff
@@ -0,0 +1,191 @@
+
+/// Represents a subject for utxos messages in the Fuel ecosystem.
+///
+/// This struct is used to create and parse subjects related to utxos messages,
+/// which can be used for subscribing to or publishing events about utxos messages.
+///
+/// # Examples
+///
+/// Creating and parsing a subject:
+///
+/// ```
+/// # use fuel_streams_core::utxos::subjects::UtxosSubject;
+/// # use fuel_streams_core::types::*;
+/// # use fuel_streams_macros::subject::*;
+/// let subject = UtxosSubject {
+///     tx_id: Some(Bytes32::from([1u8; 32]).into()),
+///     utxo_type: Some(UtxoType::Message),
+///     nonce: Some(Nonce::zeroed()),
+///     amount: Some(100),
+///     sender: Some(Address::from([2u8; 32]).into()),
+///     data: Some("".to_string()),
+///     recipient: Some(Address::from([3u8; 32]).into()),
+///     hash: None,
+/// };
+/// assert_eq!(
+///     subject.parse(),
+///     "utxos.message.*"
+/// );
+/// ```
+///
+/// All utxos messages wildcard:
+///
+/// ```
+/// # use fuel_streams_core::utxos::subjects::UtxosSubject;
+/// # use fuel_streams_macros::subject::*;
+/// assert_eq!(UtxosSubject::WILDCARD, "utxos.>");
+/// ```
+///
+/// Creating a subject query using the `wildcard` method:
+///
+/// ```
+/// # use fuel_streams_core::utxos::subjects::UtxosSubject;
+/// # use fuel_streams_core::types::*;
+/// # use fuel_streams_macros::subject::*;
+/// let wildcard = UtxosSubject::wildcard(
+///     None,
+///     Some(Bytes32::from([1u8; 32]).into()),
+///     Some(UtxoType::Message),
+///     Some(Address::from([2u8; 32]).into()),
+///     Some(Address::from([3u8; 32]).into()),
+///     Some(Nonce::zeroed()),
+///     Some("".to_string()),
+///     Some(100),
+/// );
+/// assert_eq!(wildcard, "utxos.message.*");
+/// ```
+///
+/// Using the builder pattern:
+///
+/// ```
+/// # use fuel_streams_core::utxos::subjects::UtxosSubject;
+/// # use fuel_streams_core::types::*;
+/// # use fuel_streams_macros::subject::*;
+/// let subject = UtxosSubject::new()
+///     .with_tx_id(Some(Bytes32::from([1u8; 32])))
+///     .with_utxo_type(Some(UtxoType::Message))
+///     .with_nonce(Some(Nonce::zeroed()))
+///     .with_amount(Some(100))
+///     .with_sender(Some(Address::from([2u8; 32])))
+///     .with_data(Some("".to_string()))
+///     .with_recipient(Some(Address::from([3u8; 32])))
+///     .with_computed_hash();
+/// assert_eq!(subject.parse(), "utxos.message.0x954b025b24f1c1ee5bdcb4065a4d283c1e088a9a82115d2576a7203492ada721");
+/// ```
+
+#[derive(Subject, Debug, Clone, Default)]
+#[subject_wildcard = "utxos.>"]
+#[subject_format = "utxos.{utxo_type}.{hash}"]
+#[allow(clippy::too_many_arguments)]
+pub struct UtxosSubject {
+    pub hash: Option<MessageId>,
+    pub tx_id: Option<Bytes32>,
+    pub utxo_type: Option<UtxoType>,
+    pub sender: Option<Address>,
+    pub recipient: Option<Address>,
+    pub nonce: Option<Nonce>,
+    pub data: Option<String>, // hexified data
+    pub amount: Option<u64>,
```

Comment:
I think these are meant to be in the `Utxo` struct, not here in the subject

Reply:
yes, good point. I have point. I have ported it over. Ty